[libc] Replace `MutexLock` with `cpp::lock_guard`

vmishelcs opened the pull request:
This PR address issue #89002.

#### Changes in this PR

* Added a simple implementation of `cpp::lock_guard` (an equivalent of `std::lock_guard`) in libc/src/__support/CPP inspired by the libstdc++ implementation
* Added tests for `cpp::lock_guard` in /libc/test/src/__support/CPP/mutex_test.cpp
* Replaced all references to `MutexLock` with `cpp::lock_guard`

vmishelcs committed:
[libc] Added lock_guard

vmishelcs committed:
[libc] Added test file

vmishelcs committed:
[libc] Changed lock_guard typename for clarity

vmishelcs committed:
[libc] Added basic lock_guard test

vmishelcs committed:
[libc] Added template argument in mutex_test.cpp

vmishelcs committed:
[libc] Improved mutex_test.cpp

vmishelcs committed:
[libc] Replace references to `MutexLock` with `cpp::lock_guard`

vmishelcs committed:
[libc] Ran `clang-format`

vmishelcs committed:
[libc] `clang-format` on mutex_test.cpp

vmishelcs committed:
[libc] More `clang-format`

vmishelcs committed:
[libc] Implemented `lock_guard` constructor for locked mutexes

vmishelcs committed:
[libc] `clang-format` on mutex.h

vmishelcs commented:
@nickdesaulniers Please take a look whenever you find time and let me know if I need to make any changes.

Also should I leave a comment in the new mutex.h as sort of a "bibliography" that I used the GCC implementation as inspiration?

Thanks for your time and your advice earlier on how to tackle this issue!

vmishelcs committed:
[libc] Formatting fix in mutex_test.cpp

vmishelcs committed:
[libc] Fixed comment in mutex_test.cpp

nickdesaulniers commented on libc/src/__support/threads/thread.cpp:
```diff
@@ -10,6 +10,7 @@
 #include "mutex.h"
 
 #include "src/__support/CPP/array.h"
+#include "src/__support/CPP/mutex.h" // For lock_guard
```
```suggestion
#include "src/__support/CPP/mutex.h" // lock_guard
```

nickdesaulniers commented on libc/src/stdlib/atexit.cpp:
```diff
@@ -7,6 +7,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "src/stdlib/atexit.h"
+#include "src/__support/CPP/mutex.h" // For lock_guard
```
```suggestion
#include "src/__support/CPP/mutex.h" // lock_guard
```

nickdesaulniers commented on libc/src/threads/linux/CndVar.h:
```diff
@@ -10,6 +10,7 @@
 #define LLVM_LIBC_SRC_THREADS_LINUX_CNDVAR_H
 
 #include "src/__support/CPP/atomic.h"
+#include "src/__support/CPP/mutex.h"      // For lock_guard
```
as a heads up, this hunk is likely to collide with #89329.

nickdesaulniers replied:
```suggestion
#include "src/__support/CPP/mutex.h"      // lock_guard
```

nickdesaulniers commented on libc/test/src/__support/CPP/mutex_test.cpp:
```diff
@@ -0,0 +1,67 @@
+using LIBC_NAMESPACE::cpp::adopt_lock;
+using LIBC_NAMESPACE::cpp::lock_guard;
+
+static const int SIGABRT = 6;
```
```suggestion
#include "include/llvm-libc-macros/signal-macros.h"
```

nickdesaulniers replied:
And will need to add `libc.include.llvm-libc-macros.signal_macros` to the `DEPENDS` for this test.

nickdesaulniers replied:
Actually, I think it would be simpler to just tracked with a simple boolean if your Mutex has been "double locked" and assert that that's the case rather than deal with signals and dest tests in this unit test.  Is that doable?

nickdesaulniers commented on libc/src/__support/CPP/mutex.h:
```diff
@@ -0,0 +1,47 @@
+  // Calls `m.lock()` upon resource acquisition.
+  explicit lock_guard(MutexType &m) : mutex(m) { mutex.lock(); }
+
+  // Acquires ownership of the mutex object `m` without attempting to lock
+  // it. The behavior is undefined if the current thread does not hold the
+  // lock on `m`.
```
```suggestion
  // Acquires ownership of the mutex object `m` without attempting to lock
  // it. The behavior is undefined if the current thread does not hold the
  // lock on `m`.  Does not call `m.lock()` upon resource acquisition.
```

nickdesaulniers commented on libc/src/__support/CPP/mutex.h:
```diff
@@ -0,0 +1,47 @@
+//===--- A simple lock_guard implementation ---------------------*- C++ -*-===//
```
```suggestion
//===--- self contained equivalent of std:: mutex.h ---------------------*- C++ -*-===//
```
(may need to adjust line length)

nickdesaulniers commented on libc/src/__support/File/dir.cpp:
```diff
@@ -27,7 +28,7 @@ ErrorOr<Dir *> Dir::open(const char *path) {
 
 ErrorOr<struct ::dirent *> Dir::read() {
-  MutexLock lock(&mutex);
+  cpp::lock_guard<Mutex> lock(mutex);
```
We're C++17; does that let us omit the `<Mutex>` part? If so please do here and throughout.  Example: https://godbolt.org/z/4c76K5Kb5

vmishelcs replied:
> We're C++17; does that let us omit the `<Mutex>` part?

I was getting warnings about template deduction, but I think that's because I hadn't explicitly set my C++ standard in VS Code. In any case, I see that engineers working on libcxx added this macro to their config file in (libcxx/include/__config)[https://github.com/llvm/llvm-project/blob/main/libcxx/include/__config]:

```C++
// There are a handful of public standard library types that are intended to
// support CTAD but don't need any explicit deduction guides to do so. This
// macro is used to mark them as such, which suppresses the
// '-Wctad-maybe-unsupported' compiler warning when CTAD is used in user code
// with these classes.
#  if _LIBCPP_STD_VER >= 17
#    ifdef _LIBCPP_COMPILER_CLANG_BASED
#      define _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(_ClassName)                                                              \
        template <class... _Tag>                                                                                       \
        [[maybe_unused]] _ClassName(typename _Tag::__allow_ctad...)->_ClassName<_Tag...>
#    else
#      define _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(ClassName)                                                               \
        template <class... _Tag>                                                                                       \
        ClassName(typename _Tag::__allow_ctad...)->ClassName<_Tag...>
#    endif
#  else
#    define _LIBCPP_CTAD_SUPPORTED_FOR_TYPE(_ClassName) static_assert(true, "")
#  endif
```

nickdesaulniers reviewed (commented):
> Also should I leave a comment in the new mutex.h as sort of a "bibliography" that I used the GCC implementation as inspiration?

If you put it in the commit description in the UI on github, then it will be retained when we squash+merge.  Do you mean libstdc++, or GCC?  Also, it's important that you didn't copy+paste the implementation.  Using it for reference is very much in the spirit of Free Software though.

> Thanks for your time and your advice earlier on how to tackle this issue!

You're a pro! Good job and thank you!

vmishelcs commented:
> If you put it in the commit description in the UI on github, then it will be retained when we squash+merge. Do you mean libstdc++, or GCC? Also, it's important that you didn't copy+paste the implementation. Using it for reference is very much in the spirit of Free Software though.

Yes, I used [this](https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/std_mutex.h) to essentially make sure my lock_guard implementation provides the same functionality. However, no copy-pasting was done in any of my code.

nickdesaulniers commented:
Yeah, then it's fine to mention in the PR description that you were inspired by libstdc++'s implementation.

vmishelcs committed:
[libc] Moved lock_guard data member to top of class definition

vmishelcs committed:
[libc] Adjusted comments related to lock_guard implementation

vmishelcs committed:
[libc] Simplified mutex_test.cpp

vmishelcs committed:
[libc] Formatting

vmishelcs committed:
[libc] Removed type arguments from `cpp::lock_guard`

nickdesaulniers commented on libc/test/src/__support/CPP/mutex_test.cpp:
```diff
@@ -12,56 +12,72 @@
-      // Sends signal 6.
-      abort();
+      double_unlocked = true;
     }
```
You can remove `{}` for single statement conditions. This is LLVM tree wide style.
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

nickdesaulniers commented on libc/src/__support/CPP/mutex.h:
```diff
@@ -0,0 +1,47 @@
+
+  ~lock_guard() { mutex.unlock(); }
+
+private:
```
If you're just deleting these, I don't think you need to switch back to `private` here.  Are you able to remove this?

```suggestion
```

nickdesaulniers reviewed (commented):
I'm very happy with the result here! Two minor style nits, then this LGTM.

vmishelcs committed:
[libc] Removed some unnecessary code

vmishelcs committed:
[libc] clang-format

vmishelcs commented:
Done and done!

vmishelcs commented:
@nickdesaulniers @gchatelet Friendly ping.

gchatelet commented on libc/test/src/__support/CPP/mutex_test.cpp:
```diff
@@ -0,0 +1,79 @@
+// 'unlock' which are required for the cpp::lock_guard class template.
+struct Mutex {
+  // Flag to show whether this mutex is locked.
+  bool locked;
+
+  // Flag to show if this mutex has been double locked.
+  bool double_locked;
+
+  // Flag to show if this mutex has been double unlocked.
+  bool double_unlocked;
+
+  Mutex() : locked(false), double_locked(false), double_unlocked(false) {}
```
If you default initialize the variables you can remove the constructor.

gchatelet reviewed (commented):
Just a little nit, otherwise LGTM.
Can you land this PR or do you need me to land it?

vmishelcs committed:
[libc] Small changes to mutex_test.cpp

vmishelcs commented:
> Just a little nit, otherwise LGTM. Can you land this PR or do you need me to land it?

Addressed your comment :)
I don't have write permissions so if you feel that my code meets LLVM standards I'll need someone to merge this PR.

gchatelet committed:
Merge branch 'main' into issue-89002-add-lock-guard

gchatelet committed:
Fix formatting

gchatelet commented:
Thx for your contribution @vmishelcs ! 🙏
